### PR TITLE
In the Neural Collaborative Filtering model, generalize to arbitrary layers and improve parameters (#474)

### PR DESCRIPTION
**Pull request by AniketNavlur**
* Generalized the model by building the layers in an array and processing through `differentiableReduce()`

* Using more descriptive names for model parameters

* Updated NeuMF example to work with the new parameter names

### CODE REVIEW

File: Models/Recommendation/NeuMF.swift [hunk 1]
```diff
@@ -26,95 +26,95 @@ public struct NeuMF: Module {
...
-        mfDim: Int,
-        mfReg: Float,
-        mlpLayerSizes: [Int],
-        mlpLayerRegs: [Float]
+        numUsers : Int,
+        numItems : Int,
+        mfFeatures : Int,
+        mfRegularization : Float,
+        mlpLayerSizes : [Int],
+        mlpRegularizations : [Float]
```
Comment (BradLarson) [inline Models/Recommendation/NeuMF.swift:61]: Prefixes on parameters, such as `mf`, generally aren't used in Swift. It isn't immediately clear what the prefix stands for in this case. My suggestion in the previous pull request was that `embeddingSize` and `regularization` might be better choices as names, but I'm open to others. `mlp` is a bit better known as an abbreviation in the field, so it's probably fine to leave as-is.

File: Models/Recommendation/NeuMF.swift [hunk 2]
```diff
@@ -26,95 +26,95 @@ public struct NeuMF: Module {
...
-        let userEmbedMf = self.mfUserEmbed(userIndices)
-        let itemEmbedMf = self.mfItemEmbed(itemIndices)
+        let userEmbedMf = mfUserEmbed(userIndices)
+        let itemEmbedMf = mfItemEmbed(itemIndices)
 
-        // Concatenate MF and MLP parts
         let mfVector = userEmbedMf * itemEmbedMf
+
         var mlpVector = userEmbedMlp.concatenated(with: itemEmbedMlp, alongAxis: -1)
+         mlpVector = mlpLayers.differentiableReduce(mlpVector){ $1($0) }
```
Comment (BradLarson) [inline Models/Recommendation/NeuMF.swift:112]:
```suggestion
        mlpVector = mlpLayers.differentiableReduce(mlpVector){ $1($0) }
```

File: Models/Recommendation/NeuMF.swift [hunk 3]
```diff
@@ -26,95 +26,95 @@ public struct NeuMF: Module {
...
-        // TODO: Extend it for n layers by using for loop
-        // Currently only for 4 layers
-        dense1 = Dense(
-            inputSize: self.mlpLayerSizes[0], outputSize: self.mlpLayerSizes[1], activation: relu)
-        dense2 = Dense(
-            inputSize: self.mlpLayerSizes[1], outputSize: self.mlpLayerSizes[2], activation: relu)
-        dense3 = Dense(
-            inputSize: self.mlpLayerSizes[2], outputSize: self.mlpLayerSizes[3], activation: relu)
-        finalDense = Dense(inputSize: (self.mlpLayerSizes[3] + self.mfDim), outputSize: 1)
+        for (inpSz, outSz) in zip(mlpLayerSizes, mlpLayerSizes[1...]) {
```
Comment (BradLarson) [inline Models/Recommendation/NeuMF.swift:88]: `inpSz` and `outSz` are a little hard to parse as names, and not what we'd typically see in Swift. Expanding them slightly to `inputSize` and `outputSize` might make this easier to follow.

File: Models/Recommendation/NeuMF.swift [hunk 4]
```diff
@@ -26,95 +26,95 @@ public struct NeuMF: Module {
...
-        let userEmbedMlp = self.mlpUserEmbed(userIndices)
-        let itemEmbedMlp = self.mlpItemEmbed(itemIndices)
+        let userEmbedMlp = mlpUserEmbed(userIndices)
+        let itemEmbedMlp = mlpItemEmbed(itemIndices)
 
         // MF part
-        let userEmbedMf = self.mfUserEmbed(userIndices)
-        let itemEmbedMf = self.mfItemEmbed(itemIndices)
+        let userEmbedMf = mfUserEmbed(userIndices)
+        let itemEmbedMf = mfItemEmbed(itemIndices)
```
Comment (BradLarson) [inline Models/Recommendation/NeuMF.swift:107]: Would it make sense to expand the abbreviation of `Embed` to `Embedding`? Also, if we're going to use acronyms like MLP as a suffix, we should probably capitalize all letters of the acronym (we don't do this for prefixes, due to them starting with lowercase letters).

File: Models/Recommendation/NeuMF.swift (hunk 1 above)
Comment (BradLarson) [reply Models/Recommendation/NeuMF.swift:61]: Looking at the paper, there's a clear distinction between the matrix factorization and multi-layer perceptron portions, so it can see why it might be useful to call those out separately. `mf` is still an uncommon acronym, though, so would we want to expand that to `matrixFactorization`? Again, I'm open to other suggestions.

Comment (AniketNavlur) [reply Models/Recommendation/NeuMF.swift:61]: I feel `matrixFactorizationRegularization` will be a too long variable name. How about `matrixRegularization` instead?

Comment (BradLarson) [reply Models/Recommendation/NeuMF.swift:61]: Those sound like reasonable naming choices, and if you want to go ahead with those and with the other small suggestions here, I'd be happy to bring this in as such.
